service/common/src/main/java/org/apache/polaris/service/events/AwsCloudWatchEventListener.java

```diff
@@ -0,0 +1,228 @@
+    }
+  }
+
+  private InputLogEvent createLogEvent(EventAndTimestamp eventAndTimestamp) {
+    return InputLogEvent.builder()
+        .message(eventAndTimestamp.event)
+        .timestamp(eventAndTimestamp.timestamp)
+        .build();
+  }
+
+  private void sendToCloudWatch(List<InputLogEvent> events) {
+    events.sort(Comparator.comparingLong(InputLogEvent::timestamp));
+
+    PutLogEventsRequest.Builder requestBuilder =
+        PutLogEventsRequest.builder()
+            .logGroupName(logGroup)
+            .logStreamName(logStream)
+            .logEvents(events);
+
+    synchronized (this) {
+      if (sequenceToken != null) {
+        requestBuilder.sequenceToken(sequenceToken);
+      }
+
+      try {
+        PutLogEventsResponse response = client.putLogEvents(requestBuilder.build());
+        sequenceToken = response.nextSequenceToken();
+      } catch (InvalidSequenceTokenException e) {
+        sequenceToken = getSequenceToken();
+        requestBuilder.sequenceToken(sequenceToken);
+        PutLogEventsResponse retryResponse = client.putLogEvents(requestBuilder.build());
+        sequenceToken = retryResponse.nextSequenceToken();
+      }
+    }
+  }
+
+  private void ensureLogGroupAndStream() {
+    try {
+      client.createLogGroup(CreateLogGroupRequest.builder().logGroupName(logGroup).build());
+    } catch (ResourceAlreadyExistsException ignored) {
+    }
+
+    try {
+      client.createLogStream(
+          CreateLogStreamRequest.builder().logGroupName(logGroup).logStreamName(logStream).build());
+    } catch (ResourceAlreadyExistsException ignored) {
+    }
+
+    sequenceToken = getSequenceToken();
+  }
+
+  private String getSequenceToken() {
+    DescribeLogStreamsResponse response =
+        client.describeLogStreams(
+            DescribeLogStreamsRequest.builder()
+                .logGroupName(logGroup)
+                .logStreamNamePrefix(logStream)
+                .build());
+
+    return response.logStreams().stream()
+        .filter(s -> logStream.equals(s.logStreamName()))
+        .map(LogStream::uploadSequenceToken)
+        .filter(Objects::nonNull)
+        .findFirst()
+        .orElse(null);
+  }
+
+  @PreDestroy
+  void shutdown() {
+    running = false;
+    if (backgroundTask != null) {
+      try {
+        backgroundTask.get(10, TimeUnit.SECONDS);
+      } catch (Exception e) {
+        LOGGER.error("Error waiting for background logging task to finish: {}", e.getMessage());
+      }
+    }
+    if (client != null) {
+      client.close();
+    }
+  }
+
+  private record EventAndTimestamp(String event, long timestamp) {}
+
+  private long getCurrentTimestamp(CallContext callContext) {
+    return callContext.getPolarisCallContext().getClock().millis();
+  }
+
+  // Event overrides below
+  @Override
+  public void onAfterCatalogCreated(AfterCatalogCreatedEvent event, CallContext callContext) {
+    try {
+      Map<String, Object> json = objectMapper.convertValue(event.catalog(), Map.class);
```

Comment:
IIUC this will result in a lot of duplicated code for each event type, which we should try to refactor away. Also, it may make sense to separate these from the actual "logic" of the listener.